velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,576 @@
+
+    auto* accumulator = value<ComplexTypeAccumulator>(group);
+
+    auto tracker = trackRowSize(group);
+
+    auto* rawSizes = mapVector->rawSizes();
+    auto* rawOffsets = mapVector->rawOffsets();
+    rows.applyToSelected([&](vector_size_t row) {
+      if (!decodedIntermediate_.isNullAt(row)) {
+        addToFinalAggregation(
+            mapKeys.get(),
+            flatValues,
+            indices[row],
+            rawSizes,
+            rawOffsets,
+            accumulator,
+            allocator_);
+      }
+    });
+  }
+
+ protected:
+  void initializeNewGroupsInternal(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto index : indices) {
+      new (groups[index] + offset_)
+          ComplexTypeAccumulator{inputType_, allocator_};
+    }
+  }
+
+  void destroyInternal(folly::Range<char**> groups) override {
+    for (auto* group : groups) {
+      if (isInitialized(group) && !isNull(group)) {
+        value<ComplexTypeAccumulator>(group)->free(*allocator_);
+      }
+    }
+  }
+
+ private:
+  vector_size_t countElements(char** groups, int32_t numGroups) const {
+    vector_size_t size = 0;
+    for (int32_t i = 0; i < numGroups; ++i) {
+      size += value<ComplexTypeAccumulator>(groups[i])->size();
+    }
+    return size;
+  }
+
+  // Combines a partial aggregation represented by the key-value pair at row in
+  // mapKeys and mapValues into groupMap.
+  FOLLY_ALWAYS_INLINE void addToFinalAggregation(
+      const BaseVector* mapKeys,
+      const FlatVector<int64_t>* flatValues,
+      vector_size_t index,
+      const vector_size_t* rawSizes,
+      const vector_size_t* rawOffsets,
+      ComplexTypeAccumulator* accumulator,
+      HashStringAllocator* allocator) {
+    auto size = rawSizes[index];
+    auto offset = rawOffsets[index];
+
+    for (int i = 0; i < size; ++i) {
+      const auto count = flatValues->valueAt(offset + i);
+      accumulator->addValueWithCount(*mapKeys, offset + i, count, allocator);
+    }
+  }
+
+  DecodedVector decodedKeys_;
+  DecodedVector decodedIntermediate_;
+  TypePtr inputType_;
+};
+} // namespace
+
+void registerModeAggregate(
+    const std::string& prefix,
+    bool withCompanionFunctions,
+    bool overwrite) {
+  std::vector<std::shared_ptr<AggregateFunctionSignature>> signatures = {
+      AggregateFunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("T")
+          .intermediateType("map(T,bigint)")
+          .argumentType("T")
+          .build(),
+  };
+
+  auto name = prefix + "mode";
+  registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType,
+          const core::QueryConfig& /*config*/) -> std::unique_ptr<Aggregate> {
+        VELOX_CHECK_EQ(
+            argTypes.size(),
+            1,
+            "{} ({}): unexpected number of arguments",
```

Comment:
2 format args but only `name`

Reply:
Good catch. I copied this check from other places, let me fix them also.

Reply:
https://github.com/facebookincubator/velox/pull/10501